[Chronoxy] iP

Chronoxy opened the pull request:
# Patrick The Star
> Be Curious. Read Widely. Try New Things. What people calls intelligence just boils down to curiosity - Aaron Swartz

Patrick The Star, Bestie of SpongeBob, is the *best* at remembering the tasks for you!
It is
- Easy to use!
- Easy to learn!!
- Easy to perfect!!!
- And what's better than it being **FREE**?! 🆓 NOTHING 👍 

What is required of you to do is simple:
1. Download it [here](https://github.com/Chronoxy/ip)
2. Launch it
3. Add what is there to do
4. Check it out from there!

**Features**:
- [x] Managing Tasks
- [ ] Managing Deadlines
- [ ] Managing Events

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
    public static void main(String[] args) {
        new Patrick(FILE_PATH).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Chronoxy committed:
Rename, Greet, Exit

Chronoxy committed:
Echo

Chronoxy committed:
Echo: Personalized messages

Chronoxy committed:
Add, List

Chronoxy committed:
Mark as Done v1
mark is done, unmark is not done

Chronoxy committed:
Mark as Done v2
unmark is completed

Chronoxy committed:
ToDo, Event, Deadline
ToDo done

Chronoxy committed:
ToDo, Deadline, Event
Deadline and Event done

Chronoxy committed:
List
Fixed list issue where it does not show the lines

Chronoxy committed:
Automated Text UI Testing
Done

Chronoxy committed:
Handle Errors

Chronoxy committed:
Delete

Chronoxy committed:
Enums

Chronoxy committed:
Add data/tasks.txt

Chronoxy committed:
Implement tasks.txt into list function

Chronoxy committed:
Patrick class: Java coding Standards

Rename methods to fit the Java coding standards

Chronoxy committed:
Update File Access

Update the usage of files

Chronoxy committed:
Update readTasks method

readTasks now properly populate the ArrayList

Chronoxy committed:
Patrick class: Save to tasks.txt

Update to save all changes to tasks.txt

Chronoxy committed:
Patrick class: Create new File

Include action to create new file if file does not exist

Chronoxy committed:
Merge branch 'branch-level-7'

Chronoxy committed:
Add DateFormatChecker Class

DateFormatChecker checks for the date format that user inputs

Chronoxy committed:
DateFormatChecker Class: Add getTimeFormat() method

Chronoxy committed:
Patrick Class: Add Date and Time format

Add Date and Time format for deadline tasks
Add Date and Time format for event tasks

Chronoxy committed:
Merge branch 'branch-level-8'

Chronoxy committed:
Update runtest.bat

Chronoxy committed:
Add Patrick Package

Move DateFormatChecker and Patrick into new Patrick Package

Chronoxy committed:
Add tasklist package

Seperate different tasks into relevant classes

Chronoxy committed:
Add Parser package

Chronoxy committed:
Add Storage package

Chronoxy committed:
Add Ui Package

Chronoxy committed:
Merge branch 'branch-moreoop'

Chronoxy committed:
Merge branch 'add-gradle-support' of https://github.com/Chronoxy/ip

Chronoxy committed:
Update runtest

Update runtest.bat, input.txt and test/data/tasks.txt

Chronoxy committed:
Update Ui.java

Minor changes

Chronoxy committed:
Update TaskList.java

Minor Changes

Chronoxy committed:
Add Gradle Support

Chronoxy committed:
Merge branch 'add-gradle-support'

Chronoxy committed:
Add JUnit test cases

Chronoxy committed:
Update Storage.java

Update Storage.java to support JUnit test cases

Chronoxy committed:
Update Tasklist package

Update to support JUnit Test cases

Chronoxy committed:
Update Ui.java

Update Ui.java to support JUnit Test cases

Chronoxy committed:
Merge branch 'add-junit'

Chronoxy committed:
Update Patrick class

Chronoxy committed:
Add JAR file support

Chronoxy committed:
Implement JavaDoc Comments

Chronoxy committed:
Storage Class: Implement JavaDoc Comments

Chronoxy committed:
TaskList and Parser class: Add Find Method

Chronoxy committed:
Merge branch 'branch-A-JavaDoc'

Chronoxy committed:
Merge branch 'branch-Level-9'

Chronoxy committed:
Add CheckStyle Files

Add checkstyle.xml and suppressions.xml

Chronoxy committed:
Complete checkStyleMain and checkStyleTest

Chronoxy committed:
Add GUI Files

Files: DialogBox.java, Launcher.java, Main.java, MainWindow.java

Chronoxy committed:
Add CSS Files

Chronoxy committed:
Add Images Files

Chronoxy committed:
Add View Files

Chronoxy committed:
Add GUI Feature

Chronoxy committed:
Merge branch 'branch-Level-10'

hweinian commented on src/main/java/patrick/tasklist/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+     * @return a response message indicating the result of the operation.
+     * @throws Parser.PatrickException if the input is invalid or missing required components.
+     */
+    public static String deadlineTask(String input) throws Parser.PatrickException {
+        String response;
+        String newInput = input.replace("deadline ", "");
+        if (newInput.isEmpty()) {
+            throw new Parser.PatrickException("Deadline Task Details cannot be empty!!");
+        } else if (!newInput.contains("/by")) {
+            throw new Parser.PatrickException("You are missing a '/by' in your details!!");
+        } else {
+            String taskDescription = newInput.substring(0, newInput.indexOf("/") - 1);
+            if (taskDescription.isEmpty()) {
+                throw new Parser.PatrickException("Deadline Task Description cannot be empty!!");
+            } else {
+                String deadline = newInput.substring(newInput.indexOf("/by")).replace("/by ", "");
+                if (deadline.isEmpty()) {
+                    throw new Parser.PatrickException("Deadline Task deadline cannot be empty!!");
+                } else if (DateFormatChecker.getDateFormat(deadline).equals("Unknown Format")) {
+                    throw new Parser.PatrickException("Your deadline format is incorrect.\n"
+                                                        + "Type 'formats' for the formats.");
+                } else {
+                    Task task = new Deadline(taskDescription, deadline);
+                    Storage.addList(task);
+                    response = Ui.showUserMsg(task.toString());
+                    try {
+                        Storage.appendToFile("\n" + task);
+                    } catch (IOException e) {
+                        response = "There is an error: " + e.getMessage();
+                    }
+                }
+                return response;
+            }
+        }
+    }
+}
```
Great job on validating extensively!

hweinian commented on src/main/java/patrick/storage/Storage.java:
```diff
@@ -0,0 +1,190 @@
+        fileWriter.close();
+    }
+
+    /**
+     * Exception class used for handling storage operation errors.
+     */
+    public static class StorageOperationException extends Exception {
+        /**
+         * Constructs a new {@code StorageOperationException} with the specified detail message.
+         *
+         * @param message the detail message.
+         */
+        public StorageOperationException(String message) {
+            super(message);
+        }
+    }
+
+    /**
+     * Exception class used for handling illegal values in the task data.
+     */
+    public static class IllegalValueException extends Exception {
+        /**
+         * Constructs a new {@code IllegalValueException} with the specified detail message.
+         *
+         * @param message should contain relevant information on the failed constraint(s).
+         */
+        public IllegalValueException(String message) {
+            super(message);
+        }
+    }
+}
```
Great job on considering different types of exceptions!

hweinian commented on src/main/java/patrick/parser/Parser.java:
```diff
@@ -0,0 +1,161 @@
+    /**
+     * Enum representing the different types of commands that can be parsed.
+     */
+    public enum Type {
+        LIST, BYE, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, ERROR, FORMATS, FIND
+    }
```
Great use of enums!

hweinian commented on src/main/java/patrick/tasklist/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+                    throw new Parser.PatrickException("Deadline Task deadline cannot be empty!!");
+                } else if (DateFormatChecker.getDateFormat(deadline).equals("Unknown Format")) {
+                    throw new Parser.PatrickException("Your deadline format is incorrect.\n"
+                                                        + "Type 'formats' for the formats.");
```
Indentation for wrapped line should be 8 spaces more than the parent line.

Chronoxy replied:
Thanks for spotting and letting me know. I will make the necessary changes 🙇‍♂️

hweinian reviewed (commented):
LGTM! It abide to the OOP principles. The thoroughness of your test cases and validations is done extensively. There was a lot of consideration in ensuring the code is robust and maintainable. Great work!

Chronoxy committed:
Add Assertions

Add assert feature to document important assumptions

Chronoxy committed:
Add Code Quality

Made Code Quality changes to improve readability

Chronoxy committed:
Merge pull request #1 from Chronoxy/branch-A-Assertions

Add Assertions

Chronoxy committed:
Merge branch 'master' into branch-A-CodeQuality

Chronoxy committed:
Merge pull request #2 from Chronoxy/branch-A-CodeQuality

Add Code Quality

Chronoxy committed:
Update Ui File

Edit regarding checkStyle feedbacks

Chronoxy committed:
Add gradle.yml

Add according to A-CI task

Chronoxy committed:
Merge branch 'branch-A-CI'

Chronoxy committed:
Add C-DetectDuplicates

Add extension on detecting duplicates

Chronoxy committed:
Merge branch 'branch-BCD-Extension'

Chronoxy committed:
Update Header Overview Comments

Update according to Textbook after feedback from email

Chronoxy committed:
Add images for GUI

Chronoxy committed:
Update Personality and GUI

Include usage of multiple Patrick pictures
Amend replies of Patrick to suit personality

Chronoxy committed:
Merge branch 'branch-A-Personality'

Chronoxy committed:
Add more JUnit Test Cases

Chronoxy committed:
Add _config.yml

Chronoxy committed:
Add layouts

Chronoxy committed:
Add layouts file

Add layouts to customize pages

Chronoxy committed:
Update README.md

Update with information on todo, deadline, event, list, mark

Chronoxy committed:
Delete docs/_config.yml.

Chronoxy committed:
Merge branch 'master' of https://github.com/Chronoxy/ip

Chronoxy committed:
Update README.md

Complete with all implemented features

Chronoxy committed:
Update files based on CheckStyle

Chronoxy committed:
Fix issue on JAR

Issue: System cannot find path specific

Chronoxy committed:
Update README.md

Remove duplicate of 'delete task' guide

Chronoxy committed:
Fix Unmark and Mark

Fix issue where `mark` and `unmark` non-positive values are allowed

Chronoxy committed:
Fix Todo, Event and Deadline tasks

Fix a bug where it will crash when todo task is added
Fix a bug where it will crash when description is empty
